### DISCUSSION TITLE
Considerations for engine monitoring device with SH-ESP32 + Engine top hat

### DISCUSSION THREAD

**Discussion by iuriaranda** [hatlabs/discussions#43] [Q&A] [answered]
Hello!

I'm trying to measure engine RPMs (among other things) with the SH-ESP32 + the engine HAT. The plan is to tap the alternator W post into a digital input of the engine top hat, as suggested in the documentation. There are two concerns I have with that approach:

1. will this alter the measurement of the engine panel tachometer? Since the tachometer is connected to the same w post in the alternator
2. will there be any potential significant current going through that wire? Since I'm using small gauge wire inside the SH-ESP32 enclosure I'm wondering if it'll be enough. I'm using the typical jumper wire that has 2.54 female connectors. Not sure of the actual gauge but probably something like 22 or 24 AWG

Also, do I need to fuse any of the engine sensor lines? The water tank level sensor and the hall effect sensor are in a closed circuit with the SH-ESP32 so they should be fine. But the engine sensors are grounded through the boat's ground, and although power is going to be provided by the engine hat 10mA constant-current source in the analog inputs, I'm wondering if I need to protect the lines against any accidental short in the engine block.

And finally, is there something obvious that I'm missing or I designed incorrectly?

Here's a rough sketch of how each of the sensors are wired to the whole assembly.

![Engine monitoring](https://user-images.githubusercontent.com/510809/217354969-46102191-36e3-42f6-9110-83fd7bb7d2c4.jpg)

Thanks!

**Comment by mairas** [comment, marked as answer]
Hi,

You've got great questions!

First, about connecting the engine hat in parallel with the standard tach gauge: that's completely fine. The engine hat has a 100kohm series input resistance, making it practically invisible to any other devices in the same circuit. (That applies only to digital inputs - if you're connecting devices to the analog inputs, only one should provide measurement current to the circuit).

In normal operation, the current through the digital input is insignificant: about 27V/100kohm = 0.3 mA for a 30V measurement signal. You're right, though, that if there would be some wire damage, the alternator W post could easily overheat a wire shorted e.g. to the engine block. An inline fuse as close to the W post would be a good safeguard against that. I have wondered myself whether I should include that in the documentation, but given that such fuses seem to be typically omitted from standard engine harnesses, I've omitted it for simplicity.

I don't think the other engine sensors require fusing. In case of a fault, they would either show an open circuit or be shorted to ground, and either case is harmless from the electrical safety point of view.

One note, though: when making galvanic connections to external grounded systems such as the engine and the alternator, you should keep the NMEA 2000 connection isolated, meaning that you should provide a separate 12V feed to the SH-ESP32. If you don't, it's probably not the end of the world, but ground loops definitely are a major source for noise.

**Reply by svbagh** [reply]
I have two additional questions "If you're connecting devices to the analog inputs, only one should provide measurement current to the circuit." How this is achieved? Is there a galvanic isolator available for NMEA2000?

**Reply by mairas** [reply]
I have a tendency of getting too technical in my explanations, so let me elaborate with a technical drawing. 😄

<img src="https://user-images.githubusercontent.com/407010/217761468-b4b20119-c7d7-478d-bda6-5b71889ce208.jpeg"  width="50%">

Here, we have a fuel tank sender $R$ and a fuel gauge. The fuel gauge measures the fuel sender resistance by pushing a constant current $I_m$ through the tank sender and measures the resulting voltage $V_1$. The voltage is proportional to the resistance: 

$R = {V_1 \over I_m}$. 

Thus, the gauge displays the voltage value on a suitable scale.

If you want to connect the fuel gauge to the Engine Hat as well, you can do that:

<img src="https://user-images.githubusercontent.com/407010/217762392-13c8c42a-6ac1-4e79-b52b-b46778a02ad1.jpeg" width="50%">

In this case, the Engine Hat sees the same voltage as the fuel gauge. (To help scale the value, you might want to check the actual tank sensor resistance by temporarily disconnecting the gauge and measuring the resistance with a multimeter.)

No, getting to the original question - what would happen if the Constant Current Source (CCS) was activated on the Engine Hat by placing the CCS jumper? See here:

<img src="https://user-images.githubusercontent.com/407010/217764257-269dab52-d49e-4e71-8f49-a8d3c2ab5f7a.jpeg" width="50%">

Both $I_e$ and $I_m$ would be fed to the tank sender, and resulting voltage would be:

$V_1 = R (I_m + I_e)$

and as a result, the indicated $R$ values would be much lower than true values.

However, if the fuel gauge would be disconnected, the Engine Hat would need to provide the measurement current:

<img src="https://user-images.githubusercontent.com/407010/217765655-00aa6380-8757-41e6-b9f3-eab79f47237a.jpeg" width="50%">

As described above, this would be achieved by placing the respective CCS jumper.

**Reply by mairas** [reply]
Regarding the NMEA 2000 isolation question: all standards-compliant NMEA 2000 devices have galvanic isolation. Meaning that they don't share a common ground with other circuits. 

The SH-ESP32 NMEA 2000 interface complies to the standard and is isolated. But of course, it being a DIY platform, things become more complicated. In many circumstances, you can power the SH-ESP32 by splicing the N2K power leads to the power connector. But this should only be done if the SH-ESP32 does not share ground with external devices (engine block, alternator input, grounded metallic tanks, serial interfaces). And since the Engine Hat is almost always connected to such devices, you always should provide separate power to the SH-ESP32 when using the EH.

**Reply by iuriaranda** [reply]
Thanks for your reply. Regarding the N2K isolation, wouldn't it be the same situation if I power the device using a dedicated line from the main DC panel, which also shares a common ground with everything else in the boat?

What if I connect only the top pin in the analog and digital inputs of the EH and leave the bottom ones disconnected? Wouldn't I break the ground loop like that? In such case I could power the device through the N2K connection, right?

**Reply by mairas** [reply]
You're right that the SH-ESP32 would not be isolated by itself. The main point here is to maintain isolation towards the NMEA 2000 network. The reason would be to protect the N2K network integrity - it transmits safety-critical information and shouldn't be compromised, even if in practice connecting one unisolated device there probably would work absolutely fine. It's just not the right place to take your chances. 🙂

I was going to say leaving the ground unconnected for the signal paths wouldn't help, but actually it would. The digital inputs have a very high input impedance and would be compliant to the N2K spec that way. It would still  create a big ground loop, but due to the high input impedance, current flowing through that loop would be small. The analog input, at least if the constant current source is enabled, would still be an issue. A common ground voltage would still be referenced via the low-impedance constant current circuit.

Finally, if you would be measuring tank fill levels, and the tank sensor wouldn't be connected to ground, that again would be fine. The isolation requirement of the system would still be fulfilled.

**Reply by iuriaranda** [reply]
> You're right that the SH-ESP32 would not be isolated by itself. The main point here is to maintain isolation towards the NMEA 2000 network. The reason would be to protect the N2K network integrity - it transmits safety-critical information and shouldn't be compromised, even if in practice connecting one unisolated device there probably would work absolutely fine. It's just not the right place to take your chances.

That makes total sense :slightly_smiling_face: 

Ok, so what if I power the SH-ESP32 from the engine battery? For context, the N2K network is powered by the house battery, although both batteries are grounded together. Would powering the SH-ESP32 from the engine battery solve the issue? I assume that the N2K network would be safe then, but won't there be a ground loop in the engine circuit? Also, what to do then with the ground wires of the analog and digital pins?

What would be the proper and fool-proof solution to this problem? To power the SH-ESP32 from an isolated transformer? Then I guess I'd need to connect both pins of the digital and analog inputs.

> Finally, if you would be measuring tank fill levels, and the tank sensor wouldn't be connected to ground, that again would be fine. The isolation requirement of the system would still be fulfilled.

I was going to say that the tank sensor I'm using is isolated, but now I'm not sure so I'll need to double-check. :thinking:
